core/drivers/imx_snvs.c

```diff
@@ -1,186 +1,52 @@
+ * Copyright (C) 2020 Pengutronix
+ * Rouven Czerwinski <entwicklung@pengutronix.de>
  */
 
-#include <initcall.h>
 #include <io.h>
 #include <drivers/imx_snvs.h>
-#include <kernel/misc.h>
-#include <kernel/tee_time.h>
-#include <kernel/time_source.h>
 #include <mm/core_memprot.h>
 #include <mm/core_mmu.h>
-#include <mpa.h>
-#include <utee_defines.h>
-#include <platform_config.h>
 #include <stdint.h>
-#include <tee/tee_cryp_utl.h>
 #include <types_ext.h>
 #include <trace.h>
 
-#define SNVS_LPLR	0x34
-#define SNVS_LPCR	0x38
-#define SNVS_LPMKCR	0x3C
-#define SNVS_LPSVCR	0x40
-#define SNVS_LPTGFCR	0x44
-#define SNVS_LPTDCR	0x48
-#define SNVS_LPSR	0x4C
-#define SNVS_LPSRTCMR	0x50
-#define SNVS_LPSRTCLR	0x54
-#define SNVS_LPTAR	0x58
-#define SNVS_LPSMCMR	0x5C
-#define SNVS_LPSMCLR	0x60
-#define SNVS_LPPGDR	0x64
-#define SNVS_LPGPR0_A	0x68
-#define SNVS_LPZMKR0	0x6C
-#define SNVS_LPCGR0_30	0x90
-#define SNVS_LPCGR0_31	0x94
-#define SNVS_LPCGR0_32	0x98
-#define SNVS_LPCGR0_33	0x9C
-#define SNVS_LPTDC2R	0xA0
-#define SNVS_LPTDSR	0xA4
-#define SNVS_LPTGF1CR	0xA8
-#define SNVS_LPTGF2CR	0xAC
-#define SNVS_LPAT1CR	0xC0
-#define SNVS_LPAT2CR	0xC4
-#define SNVS_LPAT3CR	0xC8
-#define SNVS_LPAT4CR	0xCC
-#define SNVS_LPAT5CR	0xD0
-#define SNVS_LPATCTLR	0xE0
-#define SNVS_LPATCLKR	0xE4
-#define SNVS_LPATRC1R	0xE8
-#define SNVS_LPATRC2R	0xEC
+#define SNVS_HPSR	0x34
 
-#define SNVS_LPCR_TOP_MASK		BIT(6)
-#define SNVS_LPCR_DP_EN_MASK		BIT(5)
-#define SNVS_LPCR_SRTC_ENV_MASK		BIT(0)
+#define HPSR_SSM_ST_MASK			GENMASK_32(11, 8)
+#define HPSR_SSM_ST_SHIFT			8
+#define HPSR_SYS_SECURITY_CFG_MASK		GENMASK_32(14, 12)
+#define HPSR_SYS_SECURITY_CFG_SHIFT		12
 
-/* SNVS LP SRTC takes 32K clock as input */
-#define SNVS_SRTC_FREQ			0x8000
-#define CNT_TO_SECS_SHIFT		15
-#define CNT_TO_SECS_MASK		GENMASK_32(14, 0)
-
-static uint64_t snvs_srtc_read_lp_counter(void)
+enum snvs_security_cfg snvs_get_security_cfg(void)
 {
 	vaddr_t snvs = core_mmu_get_va(SNVS_BASE, MEM_AREA_IO_SEC);
-	uint64_t val1, val2;
-	uint32_t val;
-
-	do {
-		val = io_read32(snvs + SNVS_LPSRTCMR);
-		val1 = val;
-		val1 <<= 32;
-		val = io_read32(snvs + SNVS_LPSRTCLR);
-		val1 |= val;
-
-		val = io_read32(snvs + SNVS_LPSRTCMR);
-		val2 = val;
-		val2 <<= 32;
-		val = io_read32(snvs + SNVS_LPSRTCLR);
-		val2 |= val;
-
-	/*
-	 * When cpu/bus runs at low freq, we may never get same value
-	 * during two consecutive read, so only compare the second value.
-	 */
-	} while ((val1 >> CNT_TO_SECS_SHIFT) != (val2 >> CNT_TO_SECS_SHIFT));
-
-	return val2;
+	uint32_t val = 0;
+
+	val = io_read32(snvs + SNVS_HPSR);
+	val &= HPSR_SYS_SECURITY_CFG_MASK;
+	val = val >> HPSR_SYS_SECURITY_CFG_SHIFT;
+	DMSG("Security CFG: 0x%01x", val);
+	if (val == 0)
+		return SNVS_SECURITY_CFG_FAB;
+	else if (val == 1)
+		return SNVS_SECURITY_CFG_OPEN;
+	else if (val > 1 && val < 4)
+		return SNVS_SECURITY_CFG_CLOSED;
+	else if (val > 4 && val < 8)
+		return SNVS_SECURITY_CFG_FIELD_RETURN;
+	return SNVS_SECURITY_CFG_OPEN;
```

Comment:
could be simplify like:
val = io_read32(snvs + SNVS_HPSR);
if (val & SNVS_HPSR_SYS_SECURITY_BAD)
   return SNVS_SECURITY_CFG_FIELD_RETURN;
else if (val & SNVS_HPSR_SYS_SECURITY_CLOSED)
   return SNVS_SECURITY_CFG_CLOSED;
else if (val & SNVS_HPSR_SYS_SECURITY_OPEN)
   return SNVS_SECURITY_CFG_OPEN;

return SNVS_SECURITY_CFG_FAB;

where:
#define SNVS_HPSR_SYS_SECURITY_BAD BIT(14)
#define SNVS_HPSR_SYS_SECURITY_CLOSED BIT(13)
#define SNVS_HPSR_SYS_SECURITY_OPEN BIT(12)

Reply:
good point, will rework, this is more readable.